cooked-validators/src/Cooked/Tx/Constraints.hs

```diff
@@ -27,78 +29,132 @@ import qualified PlutusTx as Pl
+          <> maybe mempty Pl.ownStakePubKeyHash stak
+      constr = Pl.singleton $ Pl.MustPayToPubKeyAddress (Pl.PaymentPubKeyHash p) stak mData v
+  toLedgerConstraint (PaysScript v outs) = (lkups, constr)
+    where
+      lkups = Pl.otherScript (Pl.validatorScript v)
+      constr = mconcat $
+        flip map outs $ \(d, val) ->
+          Pl.mustPayToOtherScript
+            (Pl.validatorHash $ Pl.validatorScript v)
+            (Pl.Datum $ Pl.toBuiltinData d)
+            val
+
+instance Constraint Constraints where
+  extractDatumStr (miscConstraints :=>: outConstraints) =
+    M.union
+      (M.unions (extractDatumStr <$> miscConstraints))
+      (M.unions (extractDatumStr <$> outConstraints))
+
+  toLedgerConstraint (miscConstraints :=>: outConstraints) =
+    (mconcat lkups, mconcat constrs)
+    where
+      (lkups, constrs) =
+        unzip $
+          (toLedgerConstraint <$> miscConstraints)
+            <> (toLedgerConstraint <$> outConstraints)
+
+-- | Look for the 'Pl.TxOut' corresponding to an output constraint
+-- 'OutConstraint' among the outputs of a transaction.
+findGeneratedTxOutput :: OutConstraint -> [Pl.TxOut] -> Maybe Pl.TxOut
+-- TODO Also check staking credentials
+findGeneratedTxOutput (PaysPKWithDatum targetPkh _mTargetStakePkh mTargetDatum targetVal) =
+  List.find aux
   where
-    constr = Pl.mustValidateIn (Pl.to t)
-toLedgerConstraint (After t) = (mempty, constr)
+    aux :: Pl.TxOut -> Bool
+    aux (Pl.TxOut (Pl.Address (Pl.PubKeyCredential pkh) _mStakeCredential) val mDatumHash) =
+      pkh == targetPkh
+        && val == targetVal
+        && mDatumHash == (Pl.datumHash . Pl.Datum . Pl.toBuiltinData <$> mTargetDatum)
+    aux _ = False
+findGeneratedTxOutput (PaysScript targetTypedValidator [(targetTypedDatum, targetVal)]) =
+  List.find aux
   where
-    constr = Pl.mustValidateIn (Pl.from t)
-toLedgerConstraint (ValidateIn r) = (mempty, Pl.mustValidateIn r)
-toLedgerConstraint (SignedBy hashes) = (mempty, foldMap (Pl.mustBeSignedBy . Pl.PaymentPubKeyHash) hashes)
+    aux :: Pl.TxOut -> Bool
+    aux (Pl.TxOut (Pl.Address (Pl.ScriptCredential validatorHash) _) val (Just datumHash)) =
+      validatorHash == Pl.validatorHash (Pl.validatorScript targetTypedValidator)
+        && val == targetVal
+        && datumHash == (Pl.datumHash . Pl.Datum . Pl.toBuiltinData $ targetTypedDatum)
+    aux _ = False
+findGeneratedTxOutput (PaysScript _ _) = const Nothing
 
--- | Converts a list of constraints into a 'LedgerConstraint'
-toLedgerConstraints :: [Constraint] -> LedgerConstraint a
-toLedgerConstraints cs = (mconcat lkups, mconcat constrs)
-  where
-    (lkups, constrs) = unzip $ map toLedgerConstraint cs
+-- | Reorders the outputs of a transaction according to the ordered list of
+-- output constraints that generate them. Fails in case of mismatch. The
+-- reordered outputs are put at the beginning of the list.
+orderTxOutputs :: MonadFail m => [OutConstraint] -> [Pl.TxOut] -> m [Pl.TxOut]
```

Comment:
Let's not use `MonadFail` but stick to a explicit failure type instead. The behavior from `pirouette` was too surprising for comfort! :sweat_smile:

Reply:
Agreed :) Would you simply go for a Maybe / Either or something more specific?

Reply:
If we have a single failure condition, `Maybe` is fine. Otherwise, `Either String` with a small description of what failed is fine.

Even better, perhaps, is the total version:
```haskell
orderTxOutputs :: [OutConstraint] -> [Pl.TxOut] -> [Pl.TxOut]
orderTxOutputs expctd given =
  let res = map outContrsintToTxOut expctd
   in res ++ (given Data.List.\\ res)
```